Cypress PSoC64 TF-M Integration

CharleyChu opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->

#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->
WIP

- Import TF-M release package
- Couple of change in release files required to work with MBedOS

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

### Documentation <!-- Required -->

<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [x] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->

----------------------------------------------------------------------------------------------------------------

Patater committed:
psoc64: Add TF-M compatibility

The CY8CKIT_064B0S2_4343W target was added as a replacement for the
prototype CY8CKIT_064S2_4343W target, but it is missing modifications
added to CY8CKIT_064S2_4343W to enable it to work with TF-M.

Make the CY8CKIT_064B0S2_4343W target TF-M compatible by addding flash
and region definitions from TF-M and by updating the
CY8CKIT_064B0S2_4343W linker script to create a flash image compatible
with TF-M.

0xc0170 commented on targets/targets.json:
```diff
@@ -6825,17 +6825,17 @@
-            "BLE"
+            "BLE",
+            "EXPERIMENTAL_API",
+	    "PSA"
```
fix the alignment here

Patater commented on features/FEATURE_EXPERIMENTAL_API/FEATURE_PSA/TARGET_TFM/VERSION.txt:
```diff
@@ -1 +1 @@
-1d1faca481c3
+eaaf44c87c63
```
Please ensure that the VERSION.txt file documenting with commit hash of TF-M the binaries were built with is correct. All TF-M binaries in Mbed OS are currently are build with the same TF-M branch, but this PR provides a binary for PSoC64 different from the others.

We may want a PSoC64-specific `VERSION.txt` as well as a "permissive binary license" (if source is unavailable).

CharleyChu replied:
I will update the version file, Please advise what algorithm is used to generate the hash

Patater replied:
The algorithm is: the first 12 characters of the git hash of the branch of TF-M you are using to make the binary

Patater reviewed (changes requested):
The commit "More changes" looks like a fixup commit for "psoc64: Add TF-M compatibility". Please combine these commits.

The commit title "tf-m: Add TF-M binaries for CYESKIT_064B0S2_4343W" seems wrong. Isn't it the "TARGET_CY8CKIT_064B0S2_4343W target we want to add support and binaries for? Maybe we can delete the commit if we don't need it.

Commit "Update TF-M release package for regression test" adds uses of `os_wrapper`, but commit "Changes required in TF-M release" later removes `os_wrapper`. It's easier to review and read the patchset if we don't revert changes made in earlier patches.

Commits "tf-m: Add TF-M binaries for CYESKIT_064B0S2_4343W", "Update to get 064B0S2 work with TF-M", "Update TF-M release package for regression test" all add or modify TF-M secure binaries for this target. It's a bit noisy history-wise. Are we modifying the binary in each commit to ensure that each commit builds and works on their own? If not, we should only update the binary when necessary to avoid patchset noise.

In each commit, whenever `tfm_s.hex` is modified, please also ensure that the `tfm_s.axf` is also updated to match, at least so that debugging with symbols remains remotely possible.

Patater commented on features/FEATURE_EXPERIMENTAL_API/FEATURE_PSA/TARGET_TFM/include/psa/crypto_extra.h:
We need to leave this in, sorry. Enrollment algorithm support (`alg2`) is necessary for Pelion to function.

CharleyChu replied:
Done

Patater commented on features/FEATURE_EXPERIMENTAL_API/FEATURE_PSA/TARGET_TFM/include/tfm_multi_core_api.h:
```diff
@@ -40,17 +40,17 @@ int32_t tfm_platform_ns_wait_for_s_cpu_ready(void);
- * \return \ref TFM_SUCCESS on success
- * \return \ref TFM_ERROR_GENERIC on error
+ * \return \ref OS_WRAPPER_SUCCESS on success
+ * \return \ref OS_WRAPPER_ERROR on error
```
We shouldn't be using os_wrapper nor its OS_WRAPPER error codes.

CharleyChu replied:
Done

Patater commented on features/FEATURE_EXPERIMENTAL_API/FEATURE_PSA/TARGET_TFM/include/tfm_platform_api.h:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2018-2020, Arm Limited. All rights reserved.
+ * Copyright (c) 2018-2019, Arm Limited. All rights reserved.
```
This copyright year shouldn't go backwards

CharleyChu replied:
Done.

Patater commented on features/FEATURE_EXPERIMENTAL_API/FEATURE_PSA/TARGET_TFM/include/tfm_platform_api.h:
```diff
@@ -65,31 +62,6 @@ enum tfm_platform_err_t tfm_platform_ioctl(tfm_platform_ioctl_req_t request,
- *          it returns TFM_PLATFORM_ERR_SYSTEM_ERROR.
- */
-enum tfm_platform_err_t
-tfm_platform_nv_counter_increment(uint32_t counter_id);
```
Why remove the NV counter service from TF-M targets? This code is shared with v8-M targets and isn't just used by PSoC 64.

CharleyChu replied:
Our code is based TF-Mv1.0, which doesn't include latest changes in TF-M master. I will revert the change.

Patater replied:
OK, waiting for the revert

CharleyChu replied:
Done

Patater commented on features/FEATURE_EXPERIMENTAL_API/FEATURE_PSA/TARGET_TFM/include/tfm_ns_mailbox.h:
```diff
@@ -85,17 +85,27 @@ bool tfm_ns_mailbox_is_msg_replied(mailbox_msg_handle_t handle);
+ * \return Return the handle of semaphore. or NULL if failed
  */
-const void *tfm_ns_mailbox_get_task_handle(void);
+void * tfm_ns_mailbox_hal_create_semaphore(void);
```
Is this change present in any public branch of TF-M or change request to TF-M?

CharleyChu replied:
No

Patater replied:
How will we ensure when Mbed OS upgrades TF-M next time we have the necessary changes here? We need a patch to https://github.com/ARMmbed/trusted-firmware-m/tree/dev/feature-dualcore to keep changes like these. Could you please raise a PR to https://github.com/ARMmbed/trusted-firmware-m/tree/dev/feature-dualcore with the patches we need on TF-M to support PSoC 64?

RaymondNgun replied:
Jaeden, the plan here is to use our binary to get this out the door. We will work post-release to synchronize what we can. There remains a general problem here of supporting different versions of TFM. In this particular instance, I believe we can revert this change but will not be able to in this timeframe.

CharleyChu replied:
Reverted.

Patater commented on features/FEATURE_EXPERIMENTAL_API/FEATURE_PSA/TARGET_TFM/include/tfm_mailbox.h:
```diff
@@ -135,7 +135,7 @@ struct mailbox_reply_t {
     struct mailbox_msg_t   msg;
     struct mailbox_reply_t reply;
-    const void             *owner;          /* Handle of the owner task of this
+    void                   *sem;            /* Handle of the semaphore of this
```
Will this change be available in upstream TF-M?

CharleyChu replied:
No

Patater replied:
I don't think we can accept changes here that will never be upstreamed to TF-M. We will update TF-M in the future and this would overwrite these changes.

RaymondNgun replied:
Jaeden, we are going to run into problems here as there are more PSA targets supported. I doubt all PSA targets will be on the same version of TFM.  E.g. the musca target is free to move along but that doesn't mean this target will move with it.

Patater replied:
Yes, this will be a problem if Mbed OS needs to support non-upstream versions of TF-M. Mbed OS can really only support upstream versions of TF-M, one version for all targets.

RaymondNgun replied:
This is not just non-upstream versions but different upstream versions.  There is TFM v1 and v1.1 today. Although the PSA API has unlikely changed, the exports likely have.  How will mbed os deal with different upstream versions?  Would we really force all PSA targets to conform to current mbedos version of TFM?  I think we need to have a discussion on how we handle all these cases.

CharleyChu replied:
Reverted

Patater commented on features/FEATURE_EXPERIMENTAL_API/FEATURE_PSA/TARGET_TFM/include/psa_manifest/sid.h:
```diff
@@ -130,6 +132,24 @@ extern "C" {
+
+/******** TFM_PSOC_CLIENT_TEST ********/
+#define PSOC_CLIENT_TEST_LVL2_SID                                  (0x00002000U)
+#define PSOC_CLIENT_TEST_LVL2_VERSION                              (1U)
```
Do we want to always include these tests in the production TF-M binary we ship with Mbed OS?

CharleyChu replied:
Those test ID only take effective when the test service is enabled. In release build, test service is not enabled.

Patater replied:
Is the binary added by this PR the release build or the test build?

RaymondNgun replied:
Release Core Config.  We shouldn't be using Debug or Regression builds.

Patater replied:
When mbed-os-tf-m-regression-tests is used to add target support to Mbed OS with the release core config (ConfigCoreIPC.cmake), it will avoid adding these extra test services to the header here (because TF-M doesn't put these services into this header because the binary doesn't have these services available).

Commit:
psoc64: Add TF-M release package

Signed-off-by: Charley Chu <haoc@cypress.com>

Commit:
psoc64: Update TF-M for MBedOS

    - Replace TF-M OS wrapper API with CMSIS API
    - Add busy loop in mailbox_raw_spin_lock()
    - Use static allocated semaphore for NS lock

Signed-off-by: Charley Chu <haoc@cypress.com>

Patater commented on features/FEATURE_EXPERIMENTAL_API/FEATURE_PSA/TARGET_TFM/VERSION.txt:
```diff
@@ -1 +1 @@
-1d1faca481c3
+36e356a5
```
This is only 8, not 12, characters long. The Cypress TF-M version should go in a new VERSION.txt file in the same folder as the binaries, rather than here. This is the version for all targets that are updated together, which PSOC 64 won't be for now.

CharleyChu replied:
Done

Patater commented on features/FEATURE_EXPERIMENTAL_API/FEATURE_PSA/TARGET_TFM/include/psa/initial_attestation.h:
```diff
@@ -46,7 +46,7 @@ extern "C" {
  * produced tokens.
  */
-#define PSA_INITIAL_ATTEST_MAX_TOKEN_SIZE (0x400)
+#define PSA_INITIAL_ATTEST_MAX_TOKEN_SIZE (0x500)
```
Does this need to change? If so, why? Unless this change is upstream, it's likely to get overwritten with the next TF-M upgrade and v1.1 just came out).

CharleyChu replied:
Reverted

Patater commented on features/FEATURE_EXPERIMENTAL_API/FEATURE_PSA/TARGET_TFM/include/psa_manifest/sid.h:
```diff
@@ -45,8 +45,6 @@ extern "C" {
 #define TFM_SP_PLATFORM_IOCTL_SID                                  (0x00000041U)
 #define TFM_SP_PLATFORM_IOCTL_VERSION                              (1U)
-#define TFM_SP_PLATFORM_NV_COUNTER_SID                             (0x00000042U)
-#define TFM_SP_PLATFORM_NV_COUNTER_VERSION                         (1U)
```
Let's keep the NV counter service in.

CharleyChu replied:
Done

Patater commented on features/FEATURE_EXPERIMENTAL_API/FEATURE_PSA/TARGET_TFM/include/psa_manifest/sid.h:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2019-2020, Arm Limited. All rights reserved.
```
Please leave this copyright as-is.

CharleyChu replied:
Done

Patater commented on features/FEATURE_EXPERIMENTAL_API/FEATURE_PSA/TARGET_TFM/include/tfm_multi_core_api.h:
```diff
@@ -40,17 +40,17 @@ int32_t tfm_platform_ns_wait_for_s_cpu_ready(void);
  *        The actual implementation depends on the use scenario.
  *
- * \return \ref TFM_SUCCESS on success
- * \return \ref TFM_ERROR_GENERIC on error
```
Don't change this file, please.

CharleyChu replied:
Reverted.

Patater commented on targets/TARGET_Cypress/TARGET_PSOC6/TARGET_CY8CKIT_064B0S2_4343W/LECENSE-2.0.txt:
```diff
@@ -0,0 +1,202 @@
+
```
Please rename this file from LECENSE-2.0.txt to LICENSE-2.0.txt

Patater commented on targets/TARGET_Cypress/TARGET_PSOC6/TARGET_CY8CKIT_064B0S2_4343W/partition/region_defs.h:
```diff
@@ -0,0 +1,211 @@
+ * This size of buffer is big enough to store an attestation
+ * token produced by initial attestation service
+ */
+#define PSA_INITIAL_ATTEST_TOKEN_MAX_SIZE   0x400
```
Should this be changed to 0x500 to match the other change (`PSA_INITIAL_ATTEST_MAX_TOKEN_SIZE`)? or not?

CharleyChu replied:
PSA_INITIAL_ATTEST_MAX_TOKEN_SIZE has been reverted back to 0x400

Patater commented on targets/targets.json:
```diff
@@ -6844,29 +6844,52 @@
+        "extra_labels_remove": [
+            "TFM",
+            "TFM_DUALCPU"
+        ]
```
Do we need to add `components_add` "CM0P_SECURE" and `components_remove` CM0P_SLEEP for the non-PSA target like before? We may also need to add to the NPSA target:
```
"hex_filename": "psoc6_02_cm0p_secure.hex",
"boot_scheme": "single_image",
"policy_file": "policy_single_CM0_CM4.json",
```

Commit:
psoc64: Disable accessing MXCRYPTO

MXCRYPTO is protected by TF-M, Accessing its registers will trigger HW fault

Ideally all Crypto operation should be re-direct to TF-M through PSA calls, but now only
TRNG is implemented (psa_hrng.c)

MBedTLS ALTs  disabled
CRC HAL  disabled
TRNG HAL  disabled, use psa_hrng instead.

Signed-off-by: Charley Chu <haoc@cypress.com>

Commit:
psoc64: Add 064B0S2 policy file.

Copy it to the directory where the application is built

Signed-off-by: Charley Chu <haoc@cypress.com>

Commit:
psoc64: Resolve compile and runtime issues

- Remove CY8CKIT_064B0S2_4343W_NPSA target
- Reserve 32KB non-secure flash for KV storage
  This is based on Devaraj's patch
     https://github.com/ARMmbed/mbed-os/pull/13243#issuecomment-657693560

- Relocate policy file to policy directory

Signed-off-by: Charley Chu <haoc@cypress.com>

Patater reviewed (dismissed):
We still have some maintainability concerns, but this is looking fine for this point in time. Thanks a lot!

moshe-shahar commented on targets/TARGET_Cypress/TARGET_PSOC6/TARGET_CY8CKIT_064B0S2_4343W/policy/policy_multi_CM0_CM4_tfm.json:
```diff
@@ -0,0 +1,219 @@
+        "group_private_key": "../keys/grp_priv_key.json",
+        "provision_device_private_key": false,
+        "device_private_key": "../keys/dev_priv_key.json",
+        "cy_auth": "../packets/cy_auth_2m_s0_sample.jwt"
```
is it should be cy_auth_2m_**b0**_sample.jwt?
```suggestion
        "cy_auth": "../packets/cy_auth_2m_b0_sample.jwt"
```

maclobdell reviewed (dismissed):
although I have still encountered some test instability issues (possibly isolated to when testing with gcc compiler), I think this is in good enough shape to merge.

Commit:
psoc64: Add 064B0S2 default keys

Signed-off-by: Charley Chu <haoc@cypress.com>

Commit:
psoc64: Removed PSOC specific test IDs

Signed-off-by: Charley Chu <haoc@cypress.com>

Commit:
Revert "psoc64: Disable accessing MXCRYPTO"

This reverts commit 3b44ac12e63b95b7eb93c0d4209b7f2ab37a677d.

Commit:
Use PSA TRNG if PSA is enabled and no TRNG device defined

Signed-off-by: Charley Chu <haoc@cypress.com>

Commit:
psoc64: Disable TRNG and MXCRYPTO on 064B0S2

Signed-off-by: Charley Chu <haoc@cypress.com>

Commit:
Move current TFM to TFM_V1_1

Named TFM V1.1 as TFM_V1_1 instead of TFM_V1.1 as TFM_V1.1 is invalid
as macro name

Signed-off-by: Charley Chu <haoc@cypress.com>

Commit:
Add TFM V1.0

Signed-off-by: Charley Chu <haoc@cypress.com>
